File: 03-puzzle/src/main/java/puzzle/heuristicas/DistanciaManhattan.java
package puzzle.heuristicas;

import puzzle.Puzzle;
import puzzle.utils.Heuristica;

public class DistanciaManhattan implements Heuristica<Puzzle> {
    @Override
    public int calcular(Puzzle inicial, Puzzle objetivo) {
        var puntoObjetivo = objetivo.getPosicionVacio();

        var suma = 0;
        for (int i = 0; i < inicial.getEstado().length(); i++) {
            if (i == inicial.getPosicionVacio()) continue;

            var x0 = i % inicial.getSize();
            var y0 = i / inicial.getSize();

            var x1 = puntoObjetivo % objetivo.getSize();
            var y1 = puntoObjetivo / objetivo.getSize();

            suma += Math.abs(x0 - x1) + Math.abs(y0 - y1);
        }

        return suma;

//        var p0 = inicial.getPosicionVacio();
//        var x0 = p0 % inicial.getSize();
//        var y0 = p0 / inicial.getSize();
//
//        var p1 = objetivo.getPosicionVacio();
//        var x1 = p1 % objetivo.getSize();
//        var y1 = p1 / objetivo.getSize();
//
//        return Math.abs(x0 - x1) + Math.abs(y0 - y1);
    }
}

File: 03-puzzle/src/main/kotlin/puzzle/heuristicas/DistanciaManhattan.kt
package puzzle.heuristicas

import puzzle.Puzzle
import puzzle.utils.Heuristica
import kotlin.math.abs

class DistanciaManhattan(objetivo: Puzzle) : Heuristica {
    private val posEnObjetivo = Array<Pair<Int, Int>>(objetivo.estado.size) { tile ->
        Pair(
            objetivo.estado.indexOf(tile) / objetivo.size,
            objetivo.estado.indexOf(tile) % objetivo.size
        )
    }

    override fun estimar(actual: Puzzle): Int {
        var diferencia = 0
        for (i in 0..<actual.estado.size) {
            val origen = actual.estado[i]
            if (origen == 0) continue
            val (objetivoX, objetivoY) = posEnObjetivo[origen]
            diferencia += abs(i / actual.size - objetivoX) + abs(i % actual.size - objetivoY)
        }
        return diferencia
    }
}
